### DISCUSSION TITLE
chapter_computational_complexity/summary/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#13] [Announcements]
# chapter_computational_complexity/summary/

Your first book to learn Data Structure And Algorithm.

https://www.hello-algo.com/chapter_computational_complexity/summary/

**Comment by hatkd** [comment]
作者你好，我发现本书中c语言的部分并没有给出实现，对于部分大一的学生不大友好，如果可以，我希望贡献我关于对于那一部分，并在别的平台发布不知道是否可以

**Reply by krahets** [reply]
哈喽，欢迎贡献 C 代码！详情请见 #15 
如果是个人非商业转载，请注明作者 Krahets 和出处链接；不接受商业转载。

**Reply by hatkd** [reply]
好的好的

**Comment by derekzhan** [comment]
文章中为什么要用\( 格式呢，看着好别扭 😁

**Reply by krahets** [reply]
Hi 你是指「」括号吗？

**Comment by zhuasdfg** [comment]
Q&A有点看不懂，回来再看

**Comment by FantasyLeo** [comment]
Q&A 中关于 “函数和方法这两个术语的区别是什么？”的回答很具有参考意义，有启发。但是“因此”后面的表述就不够恰当，有点误导。

**Reply by krahets** [reply]
谢谢指出！已更新

**Comment by nonhana** [comment]
争取每天浏览一整章~

**Comment by Hu-Said** [comment]
Q & A中关于自动优化尾递归的编程语言补充一下，在C# 7.0及以上版本，编译器已经开始支持尾递归的优化，称为"尾调用"。当编译器检测到一个函数调用是尾递归时，它会自动将尾递归优化为迭代形式，避免创建新的栈帧和堆栈溢出。但需要注意的是，这个优化只会在Release模式下生效，而在Debug模式下，编译器不会进行尾调用优化。

**Comment by coderguai** [comment]
每天学一点点😁

**Comment by AceLi12138** [comment]
赞赞赞

**Comment by giftfm** [comment]
都说要学习c++的特性，不过要怎么学习，怎么才算学会c++的特性？

**Comment by fantiyun** [comment]
赞，请大家继续加油！

**Comment by bclnb666** [comment]
加油

**Comment by LeoooKang** [comment]
我对函数渐进上界的定义有些困惑，到底是O(n)是函数渐进上界，还是O( f(n) )是函数渐进上界呀？

**Reply by krahets** [reply]
都是呀，你举的两个例子 `f(n) = n` ，计算时间复杂度其实就是找这个 `f(n)`

**Comment by LeoooKang** [comment]
在我看来算法效率评估中的复杂度分析与实际测试有点像加速度与速度的关系。就像看一辆车的速度性能，是看它跑一百米需要多长时间，还是看它加速到100km/h需要多长时间。
这也很像公司招人，是该看他目前的成绩，还是该看他进公司后能为公司创造的成绩。
前者是静态的，后者是动态的。

**Comment by javamyth** [comment]
这里应该得加上Go,Go 也是支持过程式编程(函数),也支持面向对象编程(方法)。 未来Go 的市场 占有也不小哦。

--------------------------------

下面以几种常见的编程语言为例来说明。

C 语言是过程式编程语言，没有面向对象的概念，所以只有函数。但我们可以通过创建结构体（struct）来模拟面向对象编程，与结构体相关联的函数就相当于其他编程语言中的方法。
Java 和 C# 是面向对象的编程语言，代码块（方法）通常作为某个类的一部分。静态方法的行为类似于函数，因为它被绑定在类上，不能访问特定的实例变量。
C++ 和 Python 既支持过程式编程（函数），也支持面向对象编程（方法）。

**Comment by javamyth** [comment]
这里建议增加支持尾递归的编程语言说明。如Lisp系列语言，函数式编程语言，GCC编译器的C语言等

-----------------------

理论上，尾递归函数的空间复杂度可以优化至 
 。不过绝大多数编程语言（例如 Java、Python、C++、Go、C# 等）不支持自动优化尾递归，因此通常认为空间复杂度是 
 。

**Comment by NgxuAnGMH** [comment]
真的特别赞 👍 太棒了这个教程

**Reply by NgxuAnGMH** [reply]
Q & A 中：函数 和 方法 两个术语的区别，初学者可以认真记一下？后续如果看书遇到这么一句话“函数是一等公民”，就可以反应过来了。在现代编程语言的设计中，得益于后端编译技术的发展，很多语言都配置了这种特性，可以简单理解为：必要时候，可以将函数作为一个参数进行传递。

**Reply by NgxuAnGMH** [reply]
如果不考虑运行时（编译器后端）的差异，其实任何一门语言的语法（编译器前端），都是特别相似的。语言之间最大的不同，还是编译技术上的创新设计。以后端为例，从 C 到 C++ 到 Java，erlang 到 go 到 Rust，每一种语言的诞生都有自己想要集中 focus 解决的问题。

**Reply by xiaoyaozizaiwoweidu** [reply]
函数是一等公民，可以赋值，可以作为参数传递，可以作为返回值

**Comment by HP-Patience** [comment]
2024.5.26打卡😊
易忘：
1.尾递归在递的操作中就已经完成数据的运算，在归的时候向上返回同样的数，所以说可以优化，但是大多数语言不支持尾递归优化，所以大多数语言的尾递归空间复杂度仍为O(n)，而不是O(1)
2.理解栈帧空间的概念，想像成子弹弹夹，递归调用函数时把函数当作子弹放入，在归的时候后进的函数先出。
3.还记得时间/空间复杂度常见类型有哪些吗？每个类型对应的例子还记得吗？

**Reply by HP-Patience** [reply]
4.循环中的变量和函数所占空间复杂度为O(1)，记得为什么吗？

**Reply by sunoyyang** [reply]
很棒

**Comment by DeBut258** [comment]
复杂度分析是评估算法效率的重要工具。在算法设计和实现过程中，时间复杂度和空间复杂度是衡量算法优劣的两个主要指标。时间复杂度用于衡量算法运行时间随输入数据规模增长的变化趋势，而空间复杂度则关注算法占用的内存空间随输入数据量增长的变化。

在时间复杂度方面主要关注的是算法在最差情况下的表现，

复杂度分析还包括平均时间复杂度和最优时间复杂度。平均时间复杂度反映算法在随机数据输入下的运行效率，最接近实际应用中的表现。最优时间复杂度则几乎不用，因为输入数据一般需要满足严格条件才能达到最佳情况。

**Reply by DeBut258** [reply]
时间复杂度O(1): 常数时间复杂度。无论输入数据规模多大，算法的执行时间始终不变。比如，访问数组中的某个元素。
```python
def get_first_element(arr):
    return arr[0]
```

**Reply by DeBut258** [reply]
时间复杂度O(log n): 对数时间复杂度。算法的执行时间随着输入数据规模的增加呈对数增长。比如，二分查找算法。
```python
def binary_search(arr, target):
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1
```

**Reply by DeBut258** [reply]
O(n): 线性时间复杂度。算法的执行时间随着输入数据规模的增加线性增长。比如，遍历一个列表。
```python
def linear_search(arr, target):
    for i in range(len(arr)):
        if arr[i] == target:
            return i
    return -1
```

**Reply by DeBut258** [reply]
O(n log n): 线性对数时间复杂度。常见于一些高效的排序算法，比如快速排序和归并排序。
```python
def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return merge(left, right)

def merge(left, right):
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result
```

**Reply by DeBut258** [reply]
O(n^2): 平方时间复杂度。常见于一些简单的排序算法，比如冒泡排序和选择排序。
```python
def bubble_sort(arr):
    n = len(arr)
    for i in range(n):
        for j in range(0, n-i-1):
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]
    return arr
```

**Comment by qi-shi-wo-men-hen-ke-xi** [comment]
第二天！

**Comment by DBook111** [comment]
一口气看到这里，打个卡

**Comment by 820566** [comment]
太牛啦

**Comment by apu-li** [comment]
看完了 稀里糊涂的 哈哈哈

**Comment by cjx20020323** [comment]
2024.11.10，干就完了

**Comment by SssFeifang** [comment]
根据【2.3.2   函数渐近上界】 中的操作数量为3+2n那块， 【2.3.3   推算方法】中完整统计，应为2+（1+1）*（5*n+1）+2*n[1+(1+1)*(n+1)] = 2+10n+2+2n+4n^2+4n = 4n^2+16n+4 = n^2+n

**Comment by Haimian377** [comment]
打卡！很清楚详细！！

**Comment by xiaqi129** [comment]
感谢作者，打卡！之前总觉得困难而逃避学习算法，坚持学习，迎难而上！

**Comment by NotUr77** [comment]
为什么大学课程 反而没有这么好的教材

**Comment by yuelsun** [comment]
打卡第二天

**Comment by Introsb** [comment]
打卡第二天

**Comment by xiaozhuge123** [comment]
day02

**Comment by KBchulan** [comment]
事实上，在开启优化选项后，c++完全可以支持尾递归优化的

**Comment by zuohd** [comment]
时间复杂度和空间复杂度分析都是遵循“留主干，剪枝蔓”原则，关注随输入数据n的增大渐变时间趋势。

**Comment by yultuz1120** [comment]
关于时间复杂度、空间复杂度这里，目前只有一个初步的轮廓了解，而对于更具体的解释觉得很抽象，后续学习一下再反过来看，标记一下

**Comment by solink7** [comment]
# 打卡 2025/6/7
还可以

**Comment by 1chuanchuan** [comment]
打卡 2025/6/9

**Comment by Ivy129** [comment]
2025.6.9打卡

**Comment by sudoo-me** [comment]
finished chapter 2. 
Suggestion: 由于对编程基础较浅的同学可能不是很清楚对象和方法，建议提供简单概念或者自学资料链接。
感谢！

**Comment by guanshangshui** [comment]
对象和方法的简单定义
1. 对象 (Object)

是什么： 对象是类的实例。你可以把类想象成一个蓝图（比如“汽车”的设计图），而对象就是根据这个蓝图制造出来的具体实物（比如“一辆红色的保时捷911”）。

代表什么： 对象代表现实世界中的一个具体事物或概念。

包含什么：

状态/属性 (State/Attributes/Properties)： 描述对象当前特征的数据。例如，一个“银行账户”对象可能有 accountNumber、accountHolderName、balance 等属性。

行为 (Behavior)： 对象能够执行的操作或功能。这些行为就是通过方法来实现的。

本质： 对象是一个数据实体。它在内存中占据空间，存储着它的属性值。

类比： 想象一个真实的物体，比如一个电饭锅。

属性： 品牌、容量、颜色、当前温度、是否在煮饭等。

行为： 煮饭、保温、取消、设置定时等（这些行为对应方法）。

2. 方法 (Method)

是什么： 方法是定义在类中的函数。它描述了对象能够执行的一个具体操作或行为。

作用：

操作对象的数据： 读取或修改对象的属性值。例如，一个 deposit(amount) 方法会增加银行账户对象的 balance 属性。

执行特定任务： 完成某个计算或功能，可能涉及到对象的内部状态。例如，一个 calculateInterest() 方法会根据账户余额和利率计算利息。

与其他对象交互： 对象之间通过调用彼此的方法来通信。

如何调用： 方法必须通过对象来调用（静态方法除外，它属于类本身）。语法通常是 对象名.方法名(参数)。例如：myAccount.deposit(100.0)。

本质： 方法是一段可执行的代码。它定义了如何操作对象的数据或执行特定的逻辑。

类比： 回到电饭锅的例子。

煮饭() 方法： 当你按下“煮饭”按钮（调用方法），电饭锅内部开始加热（执行代码），改变了它的“当前温度”和“是否在煮饭”状态（修改属性）。

查询剩余时间() 方法： 调用这个方法会返回一个时间值（读取属性或进行计算）。

**Comment by sunny20010115** [comment]
打卡第一天

**Comment by yangkegithub** [comment]
我也来打个卡，2025/9/21，真正投入算法学习，第二天

**Comment by luob8110-afk** [comment]
我也打个卡吧，正式学是算法的第二天，2025.9.26

**Comment by huilizaiyi** [comment]
第二天

**Comment by rikka060921** [comment]
20251105千里之行始于足下第一次打卡

**Comment by jyc-start** [comment]
打卡！狠狠滴学

**Comment by smile-1973** [comment]
打开 2026-1-26

**Comment by suannaimiao** [comment]
最后一个Q 缺少 ：

**Comment by NishimiyaXSean** [comment]
一口气看完第二章

**Comment by Ke-Albert** [comment]
day day up!

**Comment by Skyainbow666** [comment]
作者厉害，但是最好，每章写的例子独立，否则后面看到的有些类，或者命名空间没有具体定义，不好理解

**Comment by meilli5** [comment]
2026.5.24

**Comment by EnochxynL** [comment]
我认为在这一章有必要先详细介绍随机存取机模型（RAM），这是整个《算法》分析的基础，相当于基本公理

**Comment by LucasLuQWQ** [comment]
2026.8.31
